### DISCUSSION TITLE
Face contours

### DISCUSSION THREAD

**Discussion by shahkashani** [vladmandic/human#281] [Q&A] [answered]
Hi friends!

Just wondering if there's a good way to essentially create an outline of the detected face?

I've tried using the `silhouette` annotation, but in many cases, it doesn't seem to cover the entire face. I've also tried using some of the `contour` points that I've found in the [code](https://github.com/vladmandic/human/blob/6bb7dd26fd21132398d47822ed54911aa95c0ea4/src/face/facemeshcoords.ts#L707-L714), but as you can see, it suffers from the same problem as the silhouette:

![out](https://user-images.githubusercontent.com/1395893/180091314-32f5e252-8fb5-4876-b94c-ba002c0784c7.png)

(The numbers are mesh indexes, which _should_ correspond to the contour, but visually don't seem to.)

Interestingly, the mesh _does_ have the information to cover the entire face (at least more of the cheek), but I can't seem to find the right indexes that work consistently for all rotations:

![out](https://user-images.githubusercontent.com/1395893/180091674-6862f957-5512-4c6f-9480-f72b711a1ab7.png)

Essentially I am trying to do something like this, but I don't want the open mouth to appear in the black overlay, so I figured I would draw a polygon around the entire detected face (hence the need for a silhouette / contour that covers the entire face):

![out](https://user-images.githubusercontent.com/1395893/180092693-307fbf67-9cee-4391-b142-3a3329e4b5f5.png)

Here is the raw image:

![face2](https://user-images.githubusercontent.com/1395893/180092543-02ae3961-e301-46ab-8a3b-1603abee65b4.png)

**Comment by vladmandic** [comment]
`contours` you saw in the code are disabled as i don't see a reason to enable them as they are 2d-only and better results can be found in already present results data.

best is to consider sets of connected points as triangles and fill them. so for each face mesh point, there are 2 other points that should be considered to form a triangle - that then covers entire face precisely and leaves areas such as open mouth as unmarked by default as its not part of any triangle.

and that is exactly how its already implemented - simply set `human.draw.options.fillPolygons = true;` and you'll see how `draw` method behaves (see example screenshot).

if you want to see how to map the triangles, see `src/draw/face.ts:drawFacePolygons()` (and note that `triangulation` point map is exported as `human.faceTriangulation`)

![image](https://user-images.githubusercontent.com/57876960/180097478-3d1c8f6d-c00b-41ef-aed9-83f162b29b8c.png)

now, it can also be done using `silhouette` in results, but then you'd have to deal with unmasking open mouth, etc.
if you want to go down that path, take a look at `src/face/mask.ts` which extracts only face silouetthe from entire input.
(your use case is basically inverse of that, you need everything but silouetthe)

and one more approach (best if you're dealing with some 3d framework) is to use uv map projections.
but i don't recommend that unless you're already familiar with such concepts (using three.js or babylonjs) - but its absolutely best if what you're trying to do is project something else instead of face (e.g. perform a face-swap or something like that)
for those examples, you can take a look at <https://github.com/vladmandic/human-motion>, specifically how its done in `src/mesh.ts:drawFace()`

**Comment by shahkashani** [comment]
Thank you so much for your reply! This is incredibly useful.

So, based on what I'm reading, `silhouette` is actually perfect for me because I actually _want_ the mouth to be filled. The ideal result would be this:

![out](https://user-images.githubusercontent.com/1395893/180187828-c00d7afe-cdc1-426d-82e6-f5011d90c039.png)

However, the issue is that `silhouette` doesn't seem to cover the entire face in some cases, for instance with the above example image. I get the following:


```javascript
const f = result.face[0];
const opt = {
  fillPolygons: true,
  useDepth: false,
  color: 'black'
};
lines(ctx, f.annotations.silhouette, opt);
```

![out](https://user-images.githubusercontent.com/1395893/180187593-c3e8f985-1cd4-46fc-88cf-ea88fa73bc0c.png)

If I use the `triangulation` point map, the entire surface is covered, but the mouth is not filled:


```javascript
const f = result.face[0];
const triangulation = human.faceTriangulation;
const opt = {
  fillPolygons: true,
  useDepth: false,
  color: 'black' 
};
for (let i = 0; i < triangulation.length / 3; i++) {
  const points = [
    triangulation[i * 3 + 0],
    triangulation[i * 3 + 1],     
    triangulation[i * 3 + 2]
  ].map((index) => f.mesh[index]);
  lines(ctx, points, opt);
}
```

![out](https://user-images.githubusercontent.com/1395893/180186891-9700c741-d4a0-4e73-bc39-0bcf79402f3f.png)

I guess I could combine the two, but I feel like I'm missing something.

**Comment by vladmandic** [comment, marked as answer]
issue with `silhouette` is that edges are really close to ears so plane it defines goes through head (z-axis), not in front of face - so when you turn your head, cheekbones or nose can project more to the left/right than position of the edge of the face.

i'd stick with `triangulation` method and augment it with either:
- extend triangulation to define triangles that cover open mouth  
  (you can see the points at <https://github.com/vladmandic/human/blob/main/assets/facemesh.png>)
- augment it with plane that covers mouth only  
  use `lipsUpperOuter` and `lipsLowerOuter` landmarks instead of `silhouette`

**Comment by shahkashani** [comment]
Got it! Thank you kindly for the explanation, I'm glad I'm not missing anything obvious. Both of those strategies are straight-forward enough, so I'll do some experimentation and see 😄
